Touch on the syntax for annotating arrays. (#674)
jspecify/jspecify: docs/docs/spec.md

-   a formal parameter type of a method or constructor, as defined in
    [JLS 8.4.1]

    > This excludes the receiver parameter but includes variadic parameters (in

>     > varargs methods). Because *components* types are also a recognized

Reviewer: ```suggestion
    > varargs methods). Because *array component* types are also a recognized
```

    > location for annotations, variadic paramters (like arrays) can have

>     > annotations on multiple components. For example, `@NonNull String

Reviewer: ```suggestion
    > annotations on the array itself and/or its component type. For example, `@NonNull String
```

>     > @Nullable [] strings` means non-null elements in a nullable array.

Reviewer: ```suggestion
    > @Nullable [] strings` means `strings` is a nullable array containing non-null elements.
```

    > Similarly, `void method(@Nullable String @NonNull ... strings)` means

>     > nullable elements in a non-null array.

Reviewer: ```suggestion
    > `strings` is a non-null array containing nullable elements.
```

-   an array component type

    > An array of nullable strings is written `@Nullable String[]`. Similarly, a
    > variadic parameter whose type is "array of nullable strings" is written
    > `@Nullable String...`.
    >
    > Additionally, whenever JSpecify recognizes an annotation at a given

>     > location, it also recognizes an annotation on *an array itself* at that

Reviewer: This confused me. I thought you were saying that whenever an annotation can go on an array's component type it can also go on the array itself, which is not true...and then I realized it wasn't what you were saying.

How can we clarify? Maybe start with "When JSpecify recognizes a type-use annotation on an array itself (because it is the type of a field, parameter, return type, etc.), that annotation comes directly before the `[]`, `[dimension]`, or `...` token. For example...."

Author: I had a flash of insight—or maybe just the _feeling_ that I'd had a flash of insight. You be the judge :)

-   an array component type

    > An array of nullable strings is written `@Nullable String[]`. A variadic
    > parameter whose type is "array of nullable strings" is written `@Nullable
    > String...`. Compare this to the syntax for a nullable array of strings,

>     > shown above in the discussion of variadic parameters.

Reviewer: It's too bad that we're putting variadic parameters first, and that we have to split these up at all.

Could the variadic note above simply point to this paragraph, and put all the examples here?

Author: Yes, that sounds better. Let me know what you think of this attempt.

    > variadic parameter whose type is "array of nullable strings" is written
    > `@Nullable String...`.
    >

>     > The ability to annotate array component types is orthogonal to the ability
>     > to annotate the array itself. The array itself can be annotated in the

Reviewer: ```suggestion
    > You can annotate array component types independently from
    > the array itself. The array itself can be annotated in the
```

>     > same cases as any other type in the same position can, albeit with

Reviewer: ```suggestion
    > same cases as any non-array type in the same position can, albeit with
```

    > different syntax. For example, a method parameter could be annotated as
    > `@NonNull String @Nullable [] strings`, which means `strings` is a

>     > nullable array containing non-null elements. Similarly, `void

Reviewer: ```suggestion
    > nullable array containing non-null elements. Similarly for variadic parameters, `void
```
